lib/Serialization/ModuleFile.cpp

```diff
@@ -2202,34 +2463,48 @@ Optional<CommentInfo> ModuleFile::getCommentForDecl(const Decl *D) const {
 
   if (!DeclCommentTable)
     return None;
-
-  if (D->isImplicit())
+  // Compute the USR.
+  llvm::SmallString<128> USRBuffer;
+  llvm::raw_svector_ostream OS(USRBuffer);
+  if (ide::printDeclUSRForModuleDoc(D, OS))
     return None;
 
-  if (auto *ED = dyn_cast<ExtensionDecl>(D)) {
-    // Compute the USR.
-    llvm::SmallString<128> USRBuffer;
-    {
-      llvm::raw_svector_ostream OS(USRBuffer);
-      if (ide::printExtensionUSR(ED, OS))
-        return None;
-    }
-     return getCommentForDeclByUSR(USRBuffer.str());
-  }
+  return getCommentForDeclByUSR(USRBuffer.str());
+}
 
-  auto *VD = dyn_cast<ValueDecl>(D);
-  if (!VD)
-    return None;
+Optional<BasicDeclLocs> ModuleFile::getBasicDeclLocsForDecl(const Decl *D) const {
+  assert(D);
 
+  // Keep these as assertions instead of early exits to ensure that we are not
+  // doing extra work.  These cases should be handled by clients of this API.
+  assert(!D->hasClangNode() &&
+         "cannot find comments for Clang decls in Swift modules");
+  assert(D->getDeclContext()->getModuleScopeContext() == FileContext &&
+         "Decl is from a different serialized file");
+  if (!BasicDeclLocsTable)
+    return None;
   // Compute the USR.
   llvm::SmallString<128> USRBuffer;
-  {
-    llvm::raw_svector_ostream OS(USRBuffer);
-    if (ide::printDeclUSR(VD, OS))
-      return None;
-  }
+  llvm::raw_svector_ostream OS(USRBuffer);
+  if (ide::printDeclUSRForModuleDoc(D, OS))
+    return None;
+  auto I = BasicDeclLocsTable->find(OS.str());
+  if (I == BasicDeclLocsTable->end())
+    return None;
+  return *I;
+}
 
-  return getCommentForDeclByUSR(USRBuffer.str());
+StringRef ModuleFile::getSourceFilePathById(unsigned Id) const {
+  assert(SourceFilePathsTable);
+  auto I = SourceFilePathsTable->find(Id);
+  assert(I != SourceFilePathsTable->end());
+  return *I;
+}
+
+uint32_t ModuleFile::getDeclUSRId(StringRef USR) const {
```

Comment:
Hmm, I guess this is future-proofing for other tables keyed by USR, so that the string doesn't get repeated in the keys of every table? It does seems like if you have an integer ID you don't need a hash table any more, though.